### PR TITLE
Make cirq.Pauli a gate

### PR DESCRIPTION
**Pull request by viathor**
First step to #1041.

### CODE REVIEW

File: cirq/ops/common_gates.py [hunk 1]
```diff
@@ -536,9 +536,9 @@ def _decompose_(self, qubits):
             yield cirq.XPowGate(global_shift=-0.25).on(q)
             return
 
-        yield Y(q)**0.25
-        yield X(q)**self._exponent
-        yield Y(q)**-0.25
+        yield YPowGate(exponent=0.25).on(q)
```
Comment (Strilanc) [inline cirq/ops/common_gates.py]: Why the switch in notation here? Avoiding a cycle?

File: cirq/ops/pauli.py [hunk 2]
```diff
@@ -93,14 +75,59 @@ def __sub__(self, other_or_shift: Union['Pauli', int]
...
-    def __str__(self):
-        return self._name
 
-    def __repr__(self):
-        return 'cirq.Pauli.{!s}'.format(self)
+class _PauliX(Pauli, common_gates.XPowGate):
+    def __init__(self, *, exponent: Union[value.Symbol, float] = 1.0):
+        super(_PauliX, self).__init__(_index=0, _name='X', exponent=exponent)
+
+    def _value_equality_values_cls_(self):
```
Comment (Strilanc) [inline cirq/ops/pauli.py]: This is very strongly coupled to an internal implementation detail of `value_equality`. I would instead re-annotate XPowGate with value equality, to override the options specified by eigen gate.

File: cirq/ops/pauli.py [hunk 3]
```diff
@@ -11,29 +11,31 @@
...
 
 if TYPE_CHECKING:
     # pylint: disable=unused-import
     from typing import Tuple
 
 
-@value.value_equality
-class Pauli:
-    """Represents the X, Y, or Z axis of the Bloch sphere."""
+class Pauli(eigen_gate.EigenGate, metaclass=abc.ABCMeta):
```
Comment (Strilanc) [inline cirq/ops/pauli.py]: I think we should significantly downgrade this class. Turns all of its methods about pairs of Paulis into private utility methods.

File: cirq/ops/pauli.py [hunk 4]
```diff
@@ -11,29 +11,31 @@
...
+    """Represents the Pauli gates.
+
+    This is an abstract class with no public subclasses. The only instances
+    of private subclasses are the X, Y, or Z Pauli gates defined below.
+    """
     X = None  # type: Pauli
```
Comment (Strilanc) [inline cirq/ops/pauli.py]: Delete the X/Y/Z attributes. Just `cirq.X`, no `cirq.Pauli.X`.

Not sure if the combined `XYZ` should stay or go.

File: cirq/ops/common_gates.py (hunk 1 above)
Comment (viathor) [reply cirq/ops/common_gates.py]: Yes, exactly.

File: cirq/ops/pauli.py (hunk 3 above)
Comment (viathor) [reply cirq/ops/pauli.py]: Done for `difference()`. The other two already have clients, e.g. in clifford_gate.py and pauli_string.py.

(We could still do this, but it probably makes sense to do so in a separate PR.)

File: cirq/ops/pauli.py (hunk 4 above)
Comment (Strilanc) [reply cirq/ops/pauli.py]: Renaming it to pauli gates is fine.

File: cirq/ops/pauli.py (hunk 3 above)
Comment (Strilanc) [reply cirq/ops/pauli.py]: Okay, we can push it to a separate PR.

File: cirq/ops/pauli.py [hunk 5]
```diff
@@ -11,73 +11,56 @@
...
+    """Represents the Pauli gates.
+
+    This is an abstract class with no public subclasses. The only instances
+    of private subclasses are the X, Y, or Z Pauli gates defined below.
+    """
+    _XYZ = None  # type: Tuple[Pauli, Pauli, Pauli]
 
-    def __init__(self, *, _index: int, _name: str) -> None:
+    @staticmethod
+    def by_index(index: int) -> 'Pauli':
```
Comment (Strilanc) [inline cirq/ops/pauli.py]: Do these gates have specialized simulation methods? It's important that X and Z go fast. If not, open an issue to resolve.

Comment (viathor) [reply cirq/ops/pauli.py]: Yes. Every object that we return here is either cirq.X, cirq.Y or cirq.Z and therefore `isinstance` of XPowGate, YPowGate or ZPowGate. All `isinstance`-based specializations for these types trigger for all Pauli gates appropriately.

I've added a unit test to verify `isinstance` behavior.

File: cirq/ops/pauli.py (hunk 4 above)
Comment (viathor) [reply cirq/ops/pauli.py]: Done.

**Pull request merged**